dockerTools: update cross compilation support of the pullImage function

DarkOnion0 opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

Add support for ARM architectures that require a variant field with the ARM version (e.g. "v8", "v7"...) and change how the cross compilation is handled in the function, migrating from parameters based cross compilation to Nixpkgs cross compilation.

This PR is a successor to https://github.com/NixOS/nixpkgs/pull/119683 that is marked as stale since Oct 22, 2021.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

DarkOnion0 committed:
dockerTools: add support for ARM variants

dockerTools: reformat file according to EditorConfig

roberth commented on pkgs/build-support/docker/default.nix:
```diff
@@ -75,6 +76,13 @@ let
   defaultArch = go.GOARCH;
 
+  # Add ARM variant support (e.g. v8, v7...)
+  defaultVariant =
+    if stdenv.hostPlatform.parsed.cpu.version != null then
```
```suggestion
  defaultVariant = arch:
    if stdenv.hostPlatform.parsed.cpu.version != null && arch == defaultArch then
```

DarkOnion0 replied:
`defaultVariant` function is used in different different scopes and some are making no sense to use the check `arch == defaultArch` since the value of the arch in these function is already defined by defaultArch (see the links below).

https://github.com/DarkOnion0/nixpkgs/blob/f818cb157252a2c7aa2faac5847b189a901501ee/pkgs/build-support/docker/default.nix#L553

https://github.com/DarkOnion0/nixpkgs/blob/f818cb157252a2c7aa2faac5847b189a901501ee/pkgs/build-support/docker/default.nix#L862

What about only adding this check in the `pullImage` function (the only function where the user can set its own arch)  with an error message if `arch == defaultArch` is not true ?

roberth replied:
> What about only adding this check in the pullImage function (the only function where the user can set its own arch)

Sounds good.

> with an error message if arch == defaultArch is not true

I don't see why this would be necessary. How would you fetch the right arm image when hostPlatform is x86?
This is a little weird, but changing `arch` is the purpose of the `arch` flag.

What I think we really want is to treat the combination of `arch` and `variant` as one parameter, so that it is either completely based on `hostPlatform`, or completely based on direct user input.

We should also consider deprecating these parameters in favor of `hostPlatform`, which is how the other functions work. Looking at its long git history, it seems that this logic was added in the form of parameters as a quick and dirty shortcut to make things work, rather than as an intentional feature.
The same control will remain possible via the normal cross compilation support.

roberth replied:
> stdenv.hostPlatform.parsed.cpu.version 

This seems to duplicate the `go.GOARM` logic, which also shows that `version` may be unset.
It seems better to derive `defaultVariant` from `GOARM`.

DarkOnion0 replied:
> This seems to duplicate the `go.GOARM` logic, which also shows that `version` may be unset.

You're right, I did this originally because the `GOARM` var doesn't support the `v8` arm version but after further reading, the `v8` variant is optional, knowing that I'll be migrating the `defaultVariant` to `go.GOARM`.

DarkOnion0 replied:
> I don't see why this would be necessary. How do you get the right arm image when hostPlatform is x86?
> It's a bit weird, but changing `arch` is the purpose of the `arch` flag.

I make this proposition to prevent the user from compiling a container for another operating system than the nixpkgs host version if they don't use the cross-compiling options built into nix (e.g. `pkgs.pkgsCross.aarch64-multiplatform`...).

But, as you proposed:

> We should also consider deprecating these parameters in favor of `hostPlatform`, which is how the other functions work. Looking at its long git history, it seems that this logic was added in the form of parameters as a quick and dirty shortcut to make things work, rather than as an intentional feature.
> The same control will remain possible via the normal cross compilation support.

I will work on deprecating these flags and implement the same logic as in the other functions

roberth commented on pkgs/build-support/docker/default.nix:
```diff
@@ -133,6 +141,7 @@ rec {
         --tmpdir=$TMPDIR \
         --override-os ${os} \
         --override-arch ${arch} \
+        --override-variant ${variant} \
```
Seems like this would produce `--override-variant \`, consuming `copy` on the following line as a variant.
I think this needs `${optionalString (variant != "") "--override-variant ${variant}"} \`, though I'd also recommend `null` instead of `""`.

DarkOnion0 replied:
Done :heavy_check_mark:

roberth commented on pkgs/build-support/docker/default.nix:
```diff
@@ -541,6 +550,7 @@ rec {
           pure = writeText "${baseName}-config.json" (builtins.toJSON {
             inherit created config;
             architecture = defaultArch;
+            variant = defaultVariant;
```
Why isn't this `inherit variant`?

DarkOnion0 replied:
Why not, but I don't think I can do this without giving the user the ability to set its own CPU variant in the `buildImage` function :man_shrugging:

roberth commented:
@ofborg test docker-tools

roberth commented:
There's an evaluation error in `nixosTests.docker-tools` on `x86_64-linux` and it seems we should run those tests on all system types:

`nixos/tests/all-tests.nix`:
```diff
-  docker-tools = handleTestOn ["x86_64-linux"] ./docker-tools.nix {};
+  docker-tools = handleTest ./docker-tools.nix {};
```

roberth commented on doc/builders/images/dockertools.section.md:
```diff
@@ -236,14 +236,14 @@ Its parameters are described in the example below:
+  finalImageTag = "stable-uclibc";
   os = "linux";
-  arch = "x86_64";
+  arch ="arm";
```
```suggestion
  arch = "arm";
```

roberth commented on doc/builders/images/dockertools.section.md:
```diff
@@ -236,14 +236,14 @@ Its parameters are described in the example below:
+  imageName = "docker.io/library/busybox";
+  imageDigest = "sha256:ef320ff10026a50cf5f0213d35537ce0041ac1d96e9b7800bafd8bc9eff6c693";
+  sha256 = "sha256-9T6tP/xnZwaTH+qkbkVuXhCk+qZGE3WBY8T+Ug4qCmc=";
+  finalImageName = "docker.io/library/busybox";
```
```suggestion

  # optional parameters
  finalImageName = "docker.io/library/busybox";
```

Would be nice too see at a glance which ones are required and which ones aren't.

DarkOnion0 committed:
dockerTools: update pullImage doc

DarkOnion0 committed:
dockerTools: update system specific flags/args

dockerTools: set os to linux only

dockerTools: remove unused import

dockerTools: remove deprecated parameters in an example

dockerTools: fix typo error

DarkOnion0 commented:
For some unknown reason, cross-compiling cannot be done on `aarch64-darwin` and `x86_64-darwin`. Nix sends me the following error message:

```
error: infinite recursion encountered
```

Should we restrict the supported OS to Linux only, or is this just a common problem when attempting to cross-build a Mac application from Linux?

<details>
<summary>
Custom file used for testing
</summary>
<p>

```nix
{ pkgs, ... }:

pkgs.pkgsCross.x86_64-darwin.dockerTools.buildLayeredImage {
  name = "test";
  tag = "0.0.1";
  contents = [ pkgs.pkgsCross.x86_64-darwin.hello ];

  created = "now";

  fromImage = pkgs.pkgsCross.x86_64-darwin.dockerTools.pullImage {
    imageName = "docker.io/library/busybox";
    imageDigest = "sha256:ef320ff10026a50cf5f0213d35537ce0041ac1d96e9b7800bafd8bc9eff6c693";
    sha256 = "sha256-9T6tP/xnZwaTH+qkbkVuXhCk+qZGE3WBY8T+Ug4qCmc=";
    finalImageName = "docker.io/library/busybox";
    finalImageTag = "stable-uclibc";
    #os = "linux";
    #arch ="arm";
    #variant = "v7";
  };

  config = {
    Env = [
      "DEBUG=false"
    ];

    WorkingDir = "/";

    Entrypoint = [ "/bin/hello" ];
  };
}
```

</p>
</details>

roberth commented:
> cannot be done on `aarch64-darwin` and

Can't find a release where `pkgsCross.x86_64-darwin.stdenv` evaluated, so that does seem pretty broken, but macOS doesn't run containers anyway. We can restrict it to linux afaic.

roberth commented:
darwin as a _build_ platform should work though.

roberth commented on pkgs/build-support/docker/examples.nix:
```diff
@@ -123,6 +123,14 @@ rec {
+    sha256 = "sha256-9T6tP/xnZwaTH+qkbkVuXhCk+qZGE3WBY8T+Ug4qCmc=";
+    finalImageName = "docker.io/library/busybox";
+    finalImageTag = "stable-uclibc";
+  }
```
Syntax error. I was willing to take a risk with you to simplify the function and potentially break other users' code, but how could I do that if I'm the only one who cares to run the tests? I mean a mistake is easily made, but this problem has been around since you opened the PR and clearly you haven't tried to test the code then or when iterating.
This signals that I'm going to have to be very strict about this PR, and I'll have to spend a lot more time on it to basically do your job. Frankly, that's not something I can free up time for. Maybe you would like to be a dockerTools maintainer? I could help with that.

DarkOnion0 replied:
Hi, I'm really sorry about that, to be honest, I'm not able to easily run the tests right now and I am still a beginner, but I know it's not the good answer.
I took the time to re-run all the tests / examples and everything is working now except the cross-compilation test that fails, but if understand well according to your comment in the file, I don't have a compatible remote builder for the moment (I will try to do it in the near future), so everythings seems okay.
Sorry again 😖

roberth replied:
I'm sorry too. I shouldn't have assumed anything about you or your work.

>  I don't have a compatible remote builder

Maybe [`nix-community`](https://matrix.to/#/#nix-community:nixos.org) can help, or if you happen to have a different combination, you could change it temporarily to use two `system` values that you can build with a remote builder.

DarkOnion0 replied:
After a bit of trouble configuring my remote builder, I could finally run all the tests successfully, even the cross-compilation one!

roberth commented on pkgs/build-support/docker/default.nix:
```diff
@@ -110,8 +117,17 @@ rec {
+    , variant ? null
     }:
 
+    assert os != null -> throw "`os` parameter is no longer needed by `pullImage` function";
+    assert arch != null -> throw "`arch` parameter is no longer needed by `pullImage` function";
+    assert variant != null -> throw "`variant` parameter is no longer needed by `pullImage` function";
```
Would be nice to add some instructions for doing a "cross fetch" so to speak.

roberth commented on doc/builders/images/dockertools.section.md:
```diff
@@ -257,20 +258,19 @@ pullImage {
-- `os`, if specified, is the operating system of the fetched image. By default it's `linux`.
-
-- `arch`, if specified, is the cpu architecture of the fetched image. By default it's `x86_64`.
+The host information (arch and cpu variant if available) will be used to fetch the correct image. The target os is set to `linux` by default and can't be changed to another os.
```
`target` has a specific meaning in the context of cross compilation, where it's only relevant when compiling a compiler.

```suggestion
The host information (arch and cpu variant if available) will be used to fetch the correct image. You can fetch images for a different platform by using the Nixpkgs cross compilation functionality.
```

I don't think we have to warn about the missing functionality if there is any.

Seems to be `GOOS` actually https://github.com/opencontainers/image-spec/blob/main/config.md#properties

DarkOnion0 replied:
Thanks for the tip, I didn't know the special meaning of that word in this context :sweat_smile:

roberth commented on pkgs/build-support/docker/default.nix:
```diff
@@ -35,6 +35,9 @@
 , writePython3
 }:
 
+# Prevent infinite recursion error to happen with pkgs.pkgsCross.[ARCH]-darwin
+assert go.GOOS == "darwin" -> throw "can't build container image for `darwin` with `dockerTools`";
```
I also get this error even with the `hello` package, so it's not specific to dockerTools.

```suggestion
```

roberth commented on pkgs/build-support/docker/default.nix:
```diff
@@ -110,8 +117,17 @@ rec {
 
+    assert os != null -> throw "`os` parameter is no longer needed by `pullImage` function";
+    assert arch != null -> throw "`arch` parameter is no longer needed by `pullImage` function";
+    assert variant != null -> throw "`variant` parameter is no longer needed by `pullImage` function";
```
I may have been too quick to dismiss the `arch` and `variant` as unnecessary, because a base image may be more general than the platform for which you want to build your own things. Evaluating an extra cross nixpkgs seems a bit disproportionate for this use case.

DarkOnion0 replied:
Well, I think that most people who are building docker image with nix already have their app tightly integrated in the nix ecosystem and so, that must evaluate a cross nixpkgs to compile their app for other platforms.
Also has you said in the original discussion about this change, the other functions don't work like that. If we revert this change, I think we should also edit the other function to streamline the experience with `dockerTools`, since the other functions require evaluating an extra cross nixpkgs to be able to pick the right arch / variant for the wanted system, so in nearly every case, if you want to cross-compile a container you will need to evaluate an extra cross nixpkgs .

DarkOnion0 replied:
Well, I think that most people who are building docker image with nix already have their app tightly integrated in the nix ecosystem and so they must evaluate a cross nixpkgs to compile their app for other platforms.
Has you said in the original discussion about this change, the others functions don't work like that. If we revert this change, I think we should also edit the others functions to streamline the experience with dockerTools, creating a lot of breaking changes.
Moreover, the others functions require evaluating an extra cross nixpkgs to be able to pick the right arch / variant on the wanted system for which the container needs to run on (the value of these parameters depend on the output of the nix golang package that itself depends on arch for which nixpkgs is being imported), so in nearly every case, if you want to cross-compile a container you will need to evaluate an extra cross nixpkgs (except if you use remote builder for each arch).
At least, doesn't this break the nix philosophy where we already have something to specify platforms specific things, does it ?

_Sorry for the long answer, but it seems necessary, even if I'm not 100% sure to have correctly understood your previous message :grimacing:_

roberth replied:
I agree that it is not an essential feature.

You can remove `variant` from the parameter list and asserts here, because it didn't exist before your pr.

For the other two, you should add to the message what else to do, and you could link to the manual.

Removing `arch` is still somewhat of an experiment. I hope we can make this simplification without causing much trouble.

DarkOnion0 replied:
Ok, I'll work on this and all the remaining comments in the PR over the weekend. I will do my best to help you and make this change as smooth as possible.

DarkOnion0 committed:
dockerTools: update doc

Co-authored-by: Robert Hensing <roberth@users.noreply.github.com>

DarkOnion0 committed:
dockerTools: remove macos restriction

Co-authored-by: Robert Hensing <roberth@users.noreply.github.com>

DarkOnion0 committed:
dockerTools: update assert rules

DarkOnion0 commented:
Shouldn't I make an entry in the release notes since we are making breaking change?

roberth commented on pkgs/build-support/docker/default.nix:
```diff
@@ -110,8 +115,27 @@ rec {
+      # This function is just here to avoid having to maintain 2 pretty big error messages that are nearly identical
+      crossCompilError = parameter:
+        ''
+          The `${parameter}` parameter formerly used in the `pullImage` function to cross compile container is now deprecated in favor of native nix cross compilation.
+          As a quick fix you may want to replace 'pkgs.dockerTools.pullImage' by 'pkgs.pkgsCross.<platform>.dockerTools.pullImage' where <platform> is a valid nixpkgs host platforms.
+          For more information, see the following links:
+            - The official nix guide about cross compilation (https://nixos.org/guides/cross-compilation.html)
+            - The `pullImage` function doc (https://nixos.org/manual/nixpkgs/stable/#ssec-pkgs-dockerTools-fetchFromRegistry)
```
```suggestion
          The `${parameter}` parameter formerly used in the `pullImage` function to cross compile container is now deprecated in favor of Nixpkgs cross compilation.
          One way to migrate is to change 'pkgs.dockerTools.pullImage' to 'pkgs.pkgsCross.<platform>.dockerTools.pullImage' where <platform> corresponds to the correct ${parameter}.
          For more information, including how to obtain `pkgs` for any combinations of parameters,
          see the following links:
            - The official nix guide about cross compilation (https://nixos.org/guides/cross-compilation.html)
            - The `pullImage` function doc (https://nixos.org/manual/nixpkgs/stable/#ssec-pkgs-dockerTools-fetchFromRegistry)
```

- native cross compilation is an oxymoron :)
- Nix itself is not aware of cross compilation
- `pkgsCross` is quite limiting, so I've expanded on that in the doc references (how to obtain `pkgs` for any combinations of parameters)

roberth commented:
A note in the release notes `nixos/doc/manual/from_md/release-notes/rl-2211.section.xml` would indeed be appreciated. Don't forget to run `./nixos/doc/manual/md-to-db.sh`.

roberth commented:
You could reuse much of the corrected deprecation error for that.

DarkOnion0 committed:
dockerTools: update pullImage error message

Co-authored-by: Robert Hensing <roberth@users.noreply.github.com>

DarkOnion0 committed:
dockerTools: add entry in the breaking change section of the 22.11 release note

DarkOnion0 committed:
Merge branch 'master' into dockerTools

DarkOnion0 committed:
doc: sync manual file after the merge made in bc0c1fbad42ecc3b2cb1c2f763148fa3eb27bd64

roberth commented on pkgs/build-support/docker/examples.nix:
```diff
@@ -123,7 +123,14 @@ rec {
+    sha256 = "sha256-9T6tP/xnZwaTH+qkbkVuXhCk+qZGE3WBY8T+Ug4qCmc=";
+    finalImageName = "docker.io/library/busybox";
+    finalImageTag = "stable-uclibc";
+  };
```
There's `nixos/tests/docker-tools-cross.nix`, which uses a remote builder to cross compile to the system where the test runs. It's a bit mind bending, because the construction of the test starts with the host system rather than the build system, which is a bit unintuitive, but it works. Sadly, ofborg can't run it, because you need an appropriate remote builder to run it yourself. The ofborg builders aren't configured that way. Hydra does run the test and it works. https://hydra.nixos.org/job/nixos/trunk-combined/nixos.tests.docker-tools-cross.x86_64-linux

DarkOnion0 replied:
@utdemir Sorry for being late, I'm a little bit busy these times. I've updated the doc according to your request (I've not yet committed my changes), but there is one part I'm not sure to fully understand in your problem.

> So, when I am on a x86_64 machine, but want to produce Docker images that work on armv7, this is not what I want to use. Because if I run pkgsCross.armv7l-hf-multiplatform.dockerTools.streamLayeredImage, the streamLayeredImage executable that will emit the Docker image won't work on my computer

Despite what you claimed, I'm able to build and run the executable emitted by `streamLayeredImage` on my x86_64 computer, using `pkgs.pkgsCross.aarch64-multiplatform` to build and fetch the executable / docker image and then copy and successfully run the produced image on my PI4. Since the line you commented (down below) don't restrict the platforms it should be runs on (at least if I'm right). They are just being used to create the manifest config file for the image.

https://github.com/NixOS/nixpkgs/blob/84cf42812e82cc2d8879a45bfba9e62a1f74dc00/pkgs/build-support/docker/stream_layered_image.py#L354-L358

<details>
<summary>
Custom file used for testing
</summary>
<p>

```nix
{ pkgs, ... }:

pkgs.pkgsCross.aarch64-multiplatform.dockerTools.streamLayeredImage {
  name = "test";
  tag = "0.0.1";
  contents = [ pkgs.pkgsCross.aarch64-multiplatform.hello ];

  created = "now";

  fromImage = pkgs.pkgsCross.aarch64-multiplatform.dockerTools.pullImage {
    imageName = "docker.io/library/busybox";
    imageDigest = "sha256:ef320ff10026a50cf5f0213d35537ce0041ac1d96e9b7800bafd8bc9eff6c693";
    sha256 = "sha256-BjGm4uS98GhigVhtgEMJ4qMayqek/+cMSrCz3F2J4UQ=";
    finalImageName = "docker.io/library/busybox";
    finalImageTag = "stable-uclibc";
  };

  config = {
    WorkingDir = "/";

    Entrypoint = [ "/bin/hello" ];
  };
}
```

</p>
</details>

utdemir commented on pkgs/build-support/docker/stream_layered_image.py:
```diff
@@ -354,7 +354,8 @@ def main():
             "architecture": conf["architecture"],
-            "os": "linux",
+            "variant": conf["variant"],
+            "os": conf["os"],
```
> Ultimately, any system should be capable of streaming any image, because there's nothing system-specific about the streaming script

I completely agree! That's why I think like `streamLayeredImage` itself shouldn't come from `pkgsCross`, its contents should. This PR embeds the `os` and `variant` from `pkgsCross` means that where `streamLayeredImage` runs should be the be the target.

roberth replied:
@utdemir That is a sufficient solution, but I think it would be nice to have a system-independent interface between
 - image contents and config
 - deployment tools

By making it system-independent, details about where the deployment tool runs don't seep into the image expressions.

Until we have that, we should document your solution.

DarkOnion0 replied:
@utdemir Sorry for being late, I'm a little bit busy these times. I've updated the doc according to your request (I've not yet committed my changes), but there is one part I'm not sure to fully understand in your problem.

> So, when I am on a x86_64 machine, but want to produce Docker images that work on armv7, this is not what I want to use. Because if I run pkgsCross.armv7l-hf-multiplatform.dockerTools.streamLayeredImage, the streamLayeredImage executable that will emit the Docker image won't work on my computer

Despite what you claimed, I'm able to build and run the executable emitted by `streamLayeredImage` on my x86_64 computer, using `pkgs.pkgsCross.aarch64-multiplatform` to build and fetch the executable / docker image and then copy and successfully run the produced image on my PI4. Since the line you commented (down below) don't restrict the platforms it should be runs on (at least if I'm right). They are just being used to create the manifest config file for the image.

https://github.com/NixOS/nixpkgs/blob/84cf42812e82cc2d8879a45bfba9e62a1f74dc00/pkgs/build-support/docker/stream_layered_image.py#L354-L358

<details>
<summary>
Custom file used for testing
</summary>
<p>

```nix
{ pkgs, ... }:

pkgs.pkgsCross.aarch64-multiplatform.dockerTools.streamLayeredImage {
  name = "test";
  tag = "0.0.1";
  contents = [ pkgs.pkgsCross.aarch64-multiplatform.hello ];

  created = "now";

  fromImage = pkgs.pkgsCross.aarch64-multiplatform.dockerTools.pullImage {
    imageName = "docker.io/library/busybox";
    imageDigest = "sha256:ef320ff10026a50cf5f0213d35537ce0041ac1d96e9b7800bafd8bc9eff6c693";
    sha256 = "sha256-BjGm4uS98GhigVhtgEMJ4qMayqek/+cMSrCz3F2J4UQ=";
    finalImageName = "docker.io/library/busybox";
    finalImageTag = "stable-uclibc";
  };

  config = {
    WorkingDir = "/";

    Entrypoint = [ "/bin/hello" ];
  };
}
```

</p>
</details>

utdemir replied:
>Despite what you claimed, I'm able to build and run the executable emitted by streamLayeredImage on my x86_64 computer, using pkgs.pkgsCross.aarch64-multiplatform to build and fetch the executable / docker image and then copy and successfully run the produced image on my PI4.

>  Since the line you commented (down below) don't restrict the platforms it should be runs on (at least if I'm right)

Perfect, if this works, I am happy!

My intuition (which is admittedly not great for cross-platform) stuff that the `streamLayeredImage` executable from `pkgsCross.someOtherPlatform` would only work on `someOtherPlatform`. If it works as you described, ignore my comment!
